File: web-app/frontend/lib/api/cities.ts
import { apiClient } from './client';

export interface City {
  country: string;
  name: string;
  latitude: number;
  longitude: number;
  timezone_name: string;
  timezone: number;
  is_usa?: boolean;
}

export interface AutocompleteResponse {
  usa: City[];
  international: City[];
  total: number;
  query: string;
}

export const citiesApi = {
  /**
   * Search cities (flat list, USA first)
   */
  search: async (query: string, limit: number = 20): Promise<City[]> => {
    const response = await apiClient.get<City[]>(`/api/v1/cities/search`, {
      params: { query, limit }
    });
    return response.data;
  },

  /**
   * Lookup city by exact name
   */
  lookup: async (cityName: string): Promise<City> => {
    const response = await apiClient.get<City>(`/api/v1/cities/lookup/${encodeURIComponent(cityName)}`);
    return response.data;
  },

  /**
   * Fast autocomplete with grouped results (USA / International)
   */
  autocomplete: async (q: string, limit: number = 10): Promise<AutocompleteResponse> => {
    const response = await apiClient.get<AutocompleteResponse>(`/api/v1/cities/autocomplete`, {
      params: { q, limit }
    });
    return response.data;
  },

  /**
   * Get city service statistics
   */
  getStats: async (): Promise<{ total_cities: number; usa_cities: number; world_cities: number; load_time_ms: number }> => {
    const response = await apiClient.get(`/api/v1/cities/stats`);
    return response.data;
  }
};

File: web-app/backend/app/api/v1/endpoints/cities.py
"""
City search and lookup endpoints - OPTIMIZED VERSION
Provides fast autocomplete with grouped results (USA / International)
"""
from fastapi import APIRouter, HTTPException, Query
from typing import List, Dict, Any
from pydantic import BaseModel
from app.services.city_service import CityService

router = APIRouter()


class CityResult(BaseModel):
    country: str
    name: str
    latitude: float
    longitude: float
    timezone_name: str
    timezone: float
    is_usa: bool = False


class AutocompleteResponse(BaseModel):
    usa: List[Dict[str, Any]]
    international: List[Dict[str, Any]]
    total: int
    query: str


@router.get("/search", response_model=List[Dict[str, Any]])
async def search_cities(
    query: str = Query(..., min_length=2, description="Search query (minimum 2 characters)"),
    limit: int = Query(20, ge=1, le=100, description="Maximum number of results")
):
    """
    Search for cities by name (flat list, USA first)
    """
    try:
        results = CityService.search_cities(query, limit)
        return results
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Search error: {str(e)}")


@router.get("/lookup/{city_name:path}", response_model=Dict[str, Any])
async def lookup_city(city_name: str):
    """
    Get detailed information for a specific city by exact name
    """
    try:
        result = CityService.get_city_details(city_name)
        if result is None:
            # Try partial match
            result = CityService.get_location(city_name)
        if result is None:
            raise HTTPException(status_code=404, detail=f"City '{city_name}' not found")
        return result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Lookup error: {str(e)}")


@router.get("/autocomplete", response_model=AutocompleteResponse)
async def autocomplete_cities(
    q: str = Query(..., min_length=1, description="Search prefix"),
    limit: int = Query(10, ge=1, le=50, description="Maximum suggestions per group")
):
    """
    Fast autocomplete with grouped results (USA / International)
    
    Returns:
        - usa: List of US city matches
        - international: List of international city matches
        - total: Total number of matches found
        - query: Original search query
    """
    try:
        result = CityService.autocomplete(q, limit)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Autocomplete error: {str(e)}")


@router.get("/stats", response_model=Dict[str, Any])
async def get_city_stats():
    """
    Get city service statistics (total cities loaded, load time, etc.)
    """
    try:
        return CityService.get_stats()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Stats error: {str(e)}")
